stdenv.mkDerivation: be less strict about check execution for cross

sternenseemann opened the pull request:
**DO NOT MERGE BEFORE BRANCH-OFF**

Instead of requiring the platforms be equal, use `isCompatible` to
determine if we can execute tests. The upside of this is that we now
can execute tests for natively cross compiled package sets like
pkgsStatic, pkgsLLVM and pkgsCross.musl64 etc.

Not sure how many edge cases there will be with this. To test this
change so far, I've compiled `pkgsLLVM.nix` and `nixStatic`.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

sternenseemann committed:
stdenv.mkDerivation: be less strict about check execution for cross

Instead of requiring the platforms be equal, use `isCompatible` to
determine if we can execute tests. The upside of this is that we now
can execute tests for natively cross compiled package sets like
pkgsStatic, pkgsLLVM and pkgsCross.musl64 etc.

sternenseemann commented:
Hm, given that libtool gives us so much trouble in the static case, this may not be a good idea until we can deal with libtool's reluctancy to link statically better. I remember a related issue to this investigated recently by @alyssais.

Ericson2314 commented:
It would be nice to finish this!

Ericson2314 commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -100,8 +100,8 @@ in
   # no package has `doCheck = true`.
-  doCheck' = doCheck && stdenv.hostPlatform == stdenv.buildPlatform;
-  doInstallCheck' = doInstallCheck && stdenv.hostPlatform == stdenv.buildPlatform;
+  doCheck' = doCheck && stdenv.hostPlatform.isCompatible stdenv.buildPlatform;
+  doInstallCheck' = doInstallCheck && stdenv.hostPlatform.isCompatible stdenv.buildPlatform;
```
TBH I am not sure why we would want the current `isCompatible` baed on this.

sternenseemann replied:
Indeed, `isCompatible` is used in two places:

* `capnproto`
* linux `stdenv`

and both seem to check for executability, so maybe we should just scrap that entirely?

sternenseemann replied:
The question is, should we redesign `isCompatible` (and semantically break the API) or create a new predicate?

r-burns replied:
Yeah, I think we should just scrap the current isCompatible definition. I don't think nixpkgs has any API guarantee, certainly not for dusty corners of stdenv like this.

The name "canExecute" better describes the directionality of the relationship anyway, i.e. (A can execute B) vs. (B can execute A). Hard to tell which one is executing which when just using "isCompatible"

sternenseemann replied:
@r-burns as though to prove your point, the only other use of `isCompatible` besides `stdenv` was using it the wrong way round :)

sternenseemann commented:
> It would be nice to finish this!

I think it is much more feasible now as well for static cross, since the `libtool` fixes Alyssa has done should remove the need for the `doCheck = false` I had to add…

sternenseemann committed:
lib/systems/parse: don't consider mode switching CPUs compatible

Since we (exclusively) use isCompatible to gauge whether platform a can
execute binaries built for platform b, mode switching CPUs are not to be
considered compatible for our purposes: Switching the mode of a CPU
usually requires a reset. At the very least we can't execute a mix of
executables for the two modes which would usually be the case in nixpkgs
where we may want to execute buildInputs for the hostPlatform in
addition to nativeBuildInputs for the buildPlatform.

sternenseemann committed:
lib.systems.elaborate: expose canExecute predicate over isCompatible

canExecute is like isCompatible, but also checks that the Kernels are
_equal_, i.e. that both platforms use the same syscall interface. This
is crucial in order to actually be able to execute binaries for the
other platform.

isCompatible is dropped, since it has changed semantically and there's
no use case left in nixpkgs.

sternenseemann commented:
Pushed my take on this now, let me know if you like it. When we've finalized it, I'll add an entry for the release notes. We should target 22.11, so we have enough time to figure out where we need to enable tests again and where we should be stricter with `canExecute`.

Tested `nixStatic` and `pkgsLLVM.nix` again, the latter fails, but I'd blame it on LLVM (need to look tomorrow properly).

sternenseemann commented:
> the latter fails, but I'd blame it on LLVM (need to look tomorrow properly).

Indeed Nix 2.3 still compiles with `pkgsLLVM`/`libcxx`, but latest stable Nix doesn't.

sternenseemann committed:
libffi: disable tests if linked statically

TCL 8.6 which dejagnu requires can't be linked statically.

sternenseemann committed:
nixos/doc/rl-2211.section.md: changes w.r.t. cross check execution

r-burns commented:
Looks like the manual is failing? Not sure what the issue is there.

Tested out a handful of packages in pkgsCross.gnu32 and they all worked. Will probably need to see what happens on cross-trunk to see if any packages regress, and I imagine it'd be easy to manually disable cross testing for those. So this looks good to me, I say let's give it a shot

sternenseemann commented:
> Looks like the manual is failing? Not sure what the issue is there.

Yeah, it's a strange error I'll have to look into still.

> Tested out a handful of packages in pkgsCross.gnu32 and they all worked. Will probably need to see what happens on cross-trunk to see if any packages regress, and I imagine it'd be easy to manually disable cross testing for those. So this looks good to me, I say let's give it a shot

I agree, it seems like this would generally be no problem. I can imagine that there'll be some things we need to disable manually again and weird issues in `pkgsStatic` probably, but I think by doing this after 22.05, we'll be able to notice and fix most regressions pretty well.

sternenseemann committed:
lib.systems: inform isCompatible users about removal

sternenseemann commented:
Okay, did the following things:

- Added an error message for users of `somePlatform.isCompatible`, didn't blow up ofborg, so I hope it'll cause no problems. If not it's easy to remove.
- Rebased on `master`, release notes now work properly
- Fixed a typo in the comment for `isCompatible`

Merging now before the manual starts having merge conflicts.
